File: android/src/main/kotlin/com/LondonX/monetization_kit/entity/ColorScheme.kt
package com.LondonX.monetization_kit.entity

import android.graphics.Color

data class ColorScheme(
    val brightness: Brightness,
    val primary: FlutterColor,
    val onPrimary: FlutterColor,
    val secondary: FlutterColor,
    val onSecondary: FlutterColor,
    val error: FlutterColor,
    val onError: FlutterColor,
    val background: FlutterColor,
    val onBackground: FlutterColor,
    val surface: FlutterColor,
    val onSurface: FlutterColor,
) {
    companion object {
        fun fromRaw(raw: Map<*, *>): ColorScheme {
            val brightness = if (raw["brightness"] == "dark") Brightness.dark else Brightness.light
            val primary = raw["primary"] as Long
            val onPrimary = raw["onPrimary"] as Long
            val secondary = raw["secondary"] as Long
            val onSecondary = raw["onSecondary"] as Long
            val error = raw["error"] as Long
            val onError = raw["onError"] as Long
            val background = raw["background"] as Long
            val onBackground = raw["onBackground"] as Long
            val surface = raw["surface"] as Long
            val onSurface = raw["onSurface"] as Long
            return ColorScheme(
                brightness,
                primary.let { FlutterColor.from(it) },
                onPrimary.let { FlutterColor.from(it) },
                secondary.let { FlutterColor.from(it) },
                onSecondary.let { FlutterColor.from(it) },
                error.let { FlutterColor.from(it) },
                onError.let { FlutterColor.from(it) },
                background.let { FlutterColor.from(it) },
                onBackground.let { FlutterColor.from(it) },
                surface.let { FlutterColor.from(it) },
                onSurface.let { FlutterColor.from(it) },
            )
        }
    }
}

enum class Brightness {
    light, dark,
}

data class FlutterColor(val a: Int, val r: Int, val g: Int, val b: Int) {
    companion object {
        fun from(v: Int): FlutterColor {
            return from(v.toUInt().toLong())
        }

        fun from(v: Long): FlutterColor {
            val a = v shr 24 and 0xFF
            val r = v shr 16 and 0xFF
            val g = v shr 8 and 0xFF
            val b = v shr 0 and 0xFF
            return FlutterColor(a.toInt(), r.toInt(), g.toInt(), b.toInt())
        }
    }

    val v: Int get() = Color.argb(a, r, g, b)
}

File: ios/Classes/entity/ColorScheme.swift
//
//  FlutterColor.swift
//  monetization_kit
//
//  Created by LondonX using ChatGPT on 2023/3/22.
//

import UIKit

enum Brightness {
    case light, dark
}

struct FlutterColor {
    let a: Int
    let r: Int
    let g: Int
    let b: Int
    
    init(a: Int, r: Int, g: Int, b: Int) {
        self.a = a
        self.r = r
        self.g = g
        self.b = b
    }
    
    init(from v: Int) {
        self.init(from: UInt32(v))
    }
    
    init(from v: UInt32) {
        let a = Int(v >> 24 & 0xFF)
        let r = Int(v >> 16 & 0xFF)
        let g = Int(v >> 8 & 0xFF)
        let b = Int(v >> 0 & 0xFF)
        self.init(a: a, r: r, g: g, b: b)
    }
    
    var v: UIColor {
        return UIColor(red: CGFloat(r)/255.0, green: CGFloat(g)/255.0, blue: CGFloat(b)/255.0, alpha: CGFloat(a)/255.0)
    }
    
    func copy(a: Int? = nil, r: Int? = nil, g: Int? = nil, b: Int? = nil) -> FlutterColor {
        return FlutterColor(a: a ?? self.a, r: r ?? self.r, g: g ?? self.g, b: b ?? self.b)
    }
}

struct ColorScheme {
    let brightness: Brightness
    let primary: FlutterColor
    let onPrimary: FlutterColor
    let secondary: FlutterColor
    let onSecondary: FlutterColor
    let error: FlutterColor
    let onError: FlutterColor
    let background: FlutterColor
    let onBackground: FlutterColor
    let surface: FlutterColor
    let onSurface: FlutterColor
    
    static func fromRaw(raw: [String:Any]) -> ColorScheme {
        let brightness = (raw["brightness"] as? String == "dark") ? Brightness.dark : Brightness.light
        let primary = FlutterColor(from: raw["primary"] as! Int)
        let onPrimary = FlutterColor(from: raw["onPrimary"] as! Int)
        let secondary = FlutterColor(from: raw["secondary"] as! Int)
        let onSecondary = FlutterColor(from: raw["onSecondary"] as! Int)
        let error = FlutterColor(from: raw["error"] as! Int)
        let onError = FlutterColor(from: raw["onError"] as! Int)
        let background = FlutterColor(from: raw["background"] as! Int)
        let onBackground = FlutterColor(from: raw["onBackground"] as! Int)
        let surface = FlutterColor(from: raw["surface"] as! Int)
        let onSurface = FlutterColor(from: raw["onSurface"] as! Int)
        
        return ColorScheme(
            brightness: brightness,
            primary: primary,
            onPrimary: onPrimary,
            secondary: secondary,
            onSecondary: onSecondary,
            error: error,
            onError: onError,
            background: background,
            onBackground: onBackground,
            surface: surface,
            onSurface: onSurface
        )
    }
}
